Feature/ccmec reweight dev

hassinin opened the pull request:
resolving the extra bin after 1.2GeV and implementing the q0 range controller

gputnam committed:
Skeleton code for CCMEC reweight calculators.

hassinin committed:
Merge branch 'NuSystematics:develop' into feature/gputnam-ccmec-reweight

Commit:
First draft of CC/MEC reweighting code.

Commit:
Draft of Valencia q0q3 reweighting code.

Commit:
adding valencia_config.fhcl and update CMakeList.txt

Commit:
use nucleon target information in GENIE event record to distinguish nn vs np for mec interactions

Commit:
modified files applying bilinear interpolation

Commit:
ValenciaInterp.ToolConfig.fcl and src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc modified

jedori0228 committed:
Converting ValenciaMECq0q3ResponseCalc into header only

jedori0228 committed:
Fix systproviders/CMakeLists.txt

jedori0228 committed:
Merge branch 'ccmec_rw' into test_PR_8

Commit:
updated src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc and src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.hh

jedori0228 committed:
Fix om ValenciaMECq0q3InterpWeighting_tool

Commit:
ValenciaMECq0q3ResponseCalc.cc modified

jedori0228 committed:
Fix ValenciaMECq0q3InterpWeighting_tool.cc

jedori0228 committed:
Adding example fhicl

jedori0228 committed:
Updating file path in example ValenciaMECq0q3InterpWeighting fhicl

jedori0228 committed:
Revert "ValenciaMECq0q3ResponseCalc.cc modified"

This reverts commit 64d6277ddae0a67c29b634c8566e05490c22cd24.

jedori0228 committed:
Revert "updated src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc and src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.hh"

This reverts commit eb588d5e9e93bf25f93dc0f8e3d99b68071f2825.

jedori0228 committed:
Merge branch 'ccmec_rw' into test_PR_8

jedori0228 committed:
Adding inline

jedori0228 committed:
Fix, return default

hassinin committed:
Merge pull request #1 from jedori0228/test_PR_8

Test pr 8

Commit:
resolving the extra bin after 1.2GeV and implementing the q0 range controller

Commit:
adding ValenciaMECq0q3InterpWeighting.PH.fcl will be used instead of the corresponding toolconfig file and adding controlling the range Q0min and Q0max where the weight will be applied and

jedori0228 commented on fcl/ValenciaMECq0q3InterpWeighting.PH.fcl:
```diff
@@ -0,0 +1,119 @@
+
+            EnuMin:       0.4
+            EnuMax:       2.5
+            EnuSnapTol:   0.005
```
These part are not included in the ToolConfig (`ValenciaMECq0q3InterpWeighting.ToolConfig.fcl`), and if you want users to include these parameters, can you also update it? Btw, it is always best to first edit "ToolConfig", then run `GenerateSystProviderConfigNuSyst` on it to create "ParameterHeader (PH)" file.

jedori0228 commented:
All look good, just made one minor comment on the ToolConfig above.

Commit:
updated ToolConfig and deleted PH and to generate PH use the updated ToolConfig. I also added a controler in the ToolConfig to choose Valencia or martini

hassinin commented:
I modified the tool config

hassinin commented:
I updated ToolConfig and deleted PH and to generate PH use the updated ToolConfig. I also added a controler in the ToolConfig to choose Valencia or martini

Commit:
modified ToolConfig file to fcl/SuSAv2_MEC_reweight.ToolConfig.fcl and modified src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc such that in the ToolConfig file you just choose the model Valencia or martini such it it automatically recognises the correspondingreweight model root files from the src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc file

hassinin commented:
modified ToolConfig file to fcl/SuSAv2_MEC_reweight.ToolConfig.fcl and modified src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc such that in the ToolConfig file you just choose the model Valencia or martini such it it automatically recognises the correspondingreweight model root files from the src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc file

jedori0228 commented:
Now the module will reweight any given MEC events to some alternative model defined by the template histogram; then does the `ValenciaMECq0q3InterpWeighting` make sense? Shouldn't `MECq0q3InterpWeighting` make more sense considering that this tool can do much more thing than a reweight to Valencia?

hassinin commented:
renamed:    src/nusystematics/responsecalculators/ValenciaMECq0q3ResponseCalc.hh -> src/nusystematics/responsecalculators/MECq0q3ResponseCalc.hh and renamed:    src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc -> src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc and renamed:    src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.hh -> src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.hh

Commit:
renamed:    src/nusystematics/responsecalculators/ValenciaMECq0q3ResponseCalc.hh -> src/nusystematics/responsecalculators/MECq0q3ResponseCalc.hh and renamed:    src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.cc -> src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc and renamed:    src/nusystematics/systproviders/ValenciaMECq0q3InterpWeighting_tool.hh -> src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.hh

jedori0228 commented on src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc:
```diff
@@ -0,0 +1,379 @@
+  SystMetaData smd;
+  SystParamHeader phdr;
+  if (ParseFhiclToolConfigurationParameter(ps,
+                                           "ValenciaMECResponse",
```
`ValenciaMECResponse` -> `MECResponse`

jedori0228 commented on src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc:
```diff
@@ -0,0 +1,379 @@
+  }
+
+  // stash manifest for SetupResponseCalculator
+  auto man = ps.get<fhicl::ParameterSet>("ValenciaMECResponse_input_manifest");
```
`ValenciaMECResponse` -> `MECResponse`

jedori0228 commented on src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc:
```diff
@@ -0,0 +1,379 @@
+
+  // stash manifest for SetupResponseCalculator
+  auto man = ps.get<fhicl::ParameterSet>("ValenciaMECResponse_input_manifest");
+  tool_options.put("ValenciaMECResponse_input_manifest", man);
```
ValenciaMECResponse -> MECResponse

jedori0228 commented on src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc:
```diff
@@ -0,0 +1,379 @@
+  std::cout << "[MECq0q3InterpWeighting] SetupResponseCalculator begin\n";
+
+  const auto manifest =
+      tool_opts.get<fhicl::ParameterSet>("ValenciaMECResponse_input_manifest");
```
`ValenciaMECResponse` -> `MECResponse`

jedori0228 commented on fcl/SuSAv2_MEC_reweight.ToolConfig.fcl:
```diff
@@ -0,0 +1,51 @@
+
+  # Define the systematic parameter (dial) for interpolation weighting
+  # central value 0, evaluate at -1, 0, +1 (can expand later)
+  ValenciaMECResponse_central_value: 0
```
`ValenciaMECResponse` -> `MECResponse`

jedori0228 commented on fcl/SuSAv2_MEC_reweight.ToolConfig.fcl:
```diff
@@ -0,0 +1,51 @@
+  # Define the systematic parameter (dial) for interpolation weighting
+  # central value 0, evaluate at -1, 0, +1 (can expand later)
+  ValenciaMECResponse_central_value: 0
+  ValenciaMECResponse_variation_descriptor: "[-1,0,1]"
```
`ValenciaMECResponse` -> `MECResponse`

jedori0228 commented on fcl/SuSAv2_MEC_reweight.ToolConfig.fcl:
```diff
@@ -0,0 +1,51 @@
+  ValenciaMECResponse_central_value: 0
+  ValenciaMECResponse_variation_descriptor: "[-1,0,1]"
+
+  ValenciaMECResponse_input_manifest: {
```
`ValenciaMECResponse` -> `MECResponse`

jedori0228 commented on fcl/SuSAv2_MEC_reweight.ToolConfig.fcl:
```diff
@@ -0,0 +1,51 @@
+ValenciaInterp: {
```
Something other than having Valencia?

jedori0228 commented on fcl/SuSAv2_MEC_reweight.ToolConfig.fcl:
```diff
@@ -0,0 +1,51 @@
+
+}
+
+syst_providers : [ValenciaInterp]
```
Something other than having Valencia?

jedori0228 reviewed (changes requested):
Comments on removing "Valencia" term from some parameters

Commit:
modified:   fcl/SuSAv2_MEC_reweight.ToolConfig.fcl
        modified:   src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc

hassinin commented:
modified:   fcl/SuSAv2_MEC_reweight.ToolConfig.fcl
            modified:   src/nusystematics/systproviders/MECq0q3InterpWeighting_tool.cc

Commit:
modifying the toolconfig for 0.05 steps instead of 0.1 ,modifying the naming of the weight files, and adding w branch filling in the src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx

hassinin commented:
modifying the toolconfig for 0.05 steps instead of 0.1 ,modifying the naming of the weight files, and adding w branch filling in the src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx

Commit:
solving the sharp peak issue

jedori0228 commented on src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx:
```diff
@@ -577,13 +600,27 @@ int main(int argc, char const *argv[]) {
+    //    W^2 = M_N^2 + 2 M_N q0 - Q^2  (q0 = energy transfer)
+    {
+      // Force single-nucleon mass for consistent comparison (no cluster mass)
+      const double MN = 0.939; // GeV
```
To be consistent with [nuisance](https://github.com/NUISANCEMC/nuisance/blob/main/src/Utils/PhysConst.h#L39), we can use `0.93827203`

jedori0228 commented on data/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
 #message(STATUS "[JSKIMDEBUG] CMAKE_SOURCE_DIR = ${CMAKE_SOURCE_DIR}")
-file(GLOB DATAFILES ${CMAKE_SOURCE_DIR}/data/*)
-install(FILES ${DATAFILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/data)
+install(DIRECTORY ${CMAKE_SOURCE_DIR}/data/ValenciaMECq0q3 
```
I think the plan is to hold the template files in `sbn_data` not in nusystematics. Please revert this change

jedori0228 commented on src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx:
```diff
@@ -577,13 +600,27 @@ int main(int argc, char const *argv[]) {
     }
+    // --- Compute W from (q0, Q2) and nucleon mass *right before Fill* ---
+    //    W^2 = M_N^2 + 2 M_N q0 - Q^2  (q0 = energy transfer)
+    {
```
I don't think we need `{...}` block here. Also, move this above `if(RunNuSyst){`

jedori0228 commented on src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx:
```diff
@@ -389,6 +390,12 @@ int main(int argc, char const *argv[]) {
+    // (This does NOT touch physics scalars like w, q0, Q2.)
+    tst.Clear();
+    // Optional: set physics sentinels to catch logic mistakes in quick scans.
+    tst.w   = -999.0;
```
Well, move this just before you fill w

jedori0228 commented on src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx:
```diff
@@ -542,7 +564,8 @@ int main(int argc, char const *argv[]) {
     tst.fsprotons_KE.clear();
     for(const auto& proton: protons){
-      double this_KE = proton->KinE();
+      const double this_KE = proton->KinE();
```
Why const?

Commit:
reverting the change in data/CMakeLists.txt, using const double MN =0.93827203 in  src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx to be consistent with nuisance

Commit:
toolconfig modifcation

jedori0228 commented on data/CMakeLists.txt:
```diff
@@ -1,3 +1,7 @@
 #message(STATUS "[JSKIMDEBUG] CMAKE_SOURCE_DIR = ${CMAKE_SOURCE_DIR}")
 file(GLOB DATAFILES ${CMAKE_SOURCE_DIR}/data/*)
 install(FILES ${DATAFILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/data)
+
```
Can you just remove additional lines? not making changes in this file at all

Commit:
updated data/CMakeLists.txt and updated src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx

Commit:
updated src/nusystematics/app/DumpConfiguredTweaksNuSyst.cxx

jedori0228 commented:
Thanks, all looks good to me. Merging.
